### PR TITLE
[SHARE-889][Feature] Custom taxonomies

### PR DESCRIPTION
**Pull request by aaxelb**
A few notes for reviewers:
* Subjects with null `central_synonym` are considered part of the central taxonomy (bepress)
* Since Subjects are mutable, they're now ShareObjects with provenance and everything. Only the application user (`system`) is allowed to modify subjects in the central taxonomy. It's kinda hacky.
* Custom taxonomies are made when accepting changes that create subjects with non-null `central_synonym`.

https://openscience.atlassian.net/browse/SHARE-889

- [x] Add `Taxonomy` model
- [x] Update `Subject`, `ThroughSubjects` models
- [x] Disambiguate subjects on `uri` or (`name` and `taxonomy`)
- [x] Link subjects to taxonomy when accepting changes
- [x] Migrations
- [x] Index subjects hierarchically
- [x] Reindex works when their subjects (or other related things) change
- [x] Create bepress taxonomy when loading subjects
- [x] Admin

### CODE REVIEW

File: bots/elasticsearch/bot.py [hunk 1]
```diff
@@ -193,7 +213,10 @@ def run(self, chunk_size=500):
             else:
                 most_recent_result = pendulum.parse(self.get_most_recently_modified())
                 logger.info('Looking for %ss that have been modified after %s', model, most_recent_result)
-                qs = model.objects.filter(date_modified__gt=most_recent_result).values_list('id', flat=True)
+                q = Q(date_modified__gt=most_recent_result)
+                if hasattr(model, 'subjects') and hasattr(model, 'subject_relations'):
+                    q = q | Q(subjects__date_modified__gt=most_recent_result) | Q(subject_relations__date_modified__gt=most_recent_result)
```
Comment (chrisseto) [inline bots/elasticsearch/bot.py]: What does the `EXPLAIN` of this query look like?

File: bots/elasticsearch/util.py [hunk 2]
```diff
@@ -202,14 +202,17 @@ def fetch_creativework(pks):
                 LEFT JOIN LATERAL (
                             SELECT array_agg(DISTINCT name) AS subjects
                             FROM (
-                                SELECT unnest(ARRAY [child.name, parent.name, grand_parent.name, great_grand_parent.name])
+                                SELECT concat_ws('/', taxonomy.name, great_grand_parent.name, grand_parent.name, parent.name, child.name)
                                 FROM share_subject AS child
+                                    JOIN share_subjecttaxonomy as taxonomy ON child.taxonomy_id = child.id
```
Comment (chrisseto) [inline bots/elasticsearch/util.py]: nitpick: `AS`

File: bots/elasticsearch/util.py [hunk 3]
```diff
@@ -202,14 +202,17 @@ def fetch_creativework(pks):
                 LEFT JOIN LATERAL (
                             SELECT array_agg(DISTINCT name) AS subjects
                             FROM (
-                                SELECT unnest(ARRAY [child.name, parent.name, grand_parent.name, great_grand_parent.name])
+                                SELECT concat_ws('/', taxonomy.name, great_grand_parent.name, grand_parent.name, parent.name, child.name)
```
Comment (chrisseto) [inline bots/elasticsearch/util.py]: Will this properly handle 1st and 2nd level subjects?

File: share/disambiguation.py [hunk 4]
```diff
@@ -144,6 +144,32 @@ def _instance_for_node(self, node):
...
+                pass
+
+        name = node.attrs.get('name')
+        if node.related('central_synonym') is None:
+            # Central taxonomy
+            if name:
+                try:
+                    return node.model.objects.get(name=name, central_synonym__isnull=True)
+                except node.model.DoesNotExist:
+                    logger.debug('No %s found with name "%s" in the central taxonomy', node.model, name)
```
Comment (chrisseto) [inline share/disambiguation.py]: Should this be an exception?

File: share/disambiguation.py [hunk 5]
```diff
@@ -144,6 +144,32 @@ def _instance_for_node(self, node):
...
+        else:
+            # Custom taxonomy
+            source = node.graph.source
+            if source and name:
+                try:
+                    return node.model.objects.get(name=name, taxonomy__name=source.long_title)
+                except node.model.DoesNotExist:
+                    logger.debug('No %s found with name "%s" in taxonomy "%s"', node.model, name, source.long_title)
```
Comment (chrisseto) [inline share/disambiguation.py]: Exception?

File: share/management/commands/disambiguate.py [hunk 6]
```diff
@@ -1,30 +1,22 @@
...
+        ids = options['normalized']
+        if options['source']:
+            ids.extend(NormalizedData.objects.filter(source__source__name=options['source']).values_list('id', flat=True))
+        for id in ids:
             if options['async']:
-                DisambiguatorTask().apply_async(task_args)
+                disambiguate.apply_async((id,))
             else:
-                DisambiguatorTask().apply(task_args, throw=True)
+                disambiguate(id)
```
Comment (chrisseto) [inline share/management/commands/disambiguate.py]: `disambiguate.apply((id, ))` will add tasks to the backend, if that would be desired.

File: share/migrations/0038_trust_system_user.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+def trust_system_user(apps, schema_editor):
+    ShareUser = apps.get_model('share', 'ShareUser')
+    Source = apps.get_model('share', 'Source')
+
+    system_user = ShareUser.objects.get(username=settings.APPLICATION_USERNAME)
+    system_user.is_trusted = True
+    system_user.save()
+
+    Source.objects.update_or_create(
```
Comment (chrisseto) [inline share/migrations/0038_trust_system_user.py]: Should we exclude the System source from all the source listing?

File: share/migrations/0040_subject_provenance.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                    '@id': '_:{}'.format(s.id),
+                    '@type': 'subject',
+                    'name': s.name,
+                    'parent': None if s.parent_id is None else {'@id': '_:{}'.format(s.parent_id), '@type': 'subject'}
+                } for s in Subject.objects.all()
+            ]
+        }
+    )
+    from share.tasks import DisambiguatorTask
+    DisambiguatorTask().apply((user.id, normalized_data.id), throw=True)
```
Comment (chrisseto) [inline share/migrations/0040_subject_provenance.py]: This needs to be updated.

File: share/models/change.py [hunk 9]
```diff
@@ -197,7 +198,8 @@ def _update(self, save=True):
         if new_type:
             self.target.recast('share.{}'.format(new_type))
 
-        self.target.__dict__.update(self._resolve_change())
+        for k, v in self._resolve_change().items():
+            setattr(self.target, k, v)
         if save:
```
Comment (chrisseto) [inline share/models/change.py]: nitpick: Add a newline

File: share/models/change.py [hunk 10]
```diff
@@ -274,4 +276,15 @@ def _resolve_change(self):
                 change[k] = [self.change_set._resolve_ref(r) for r in v]
             else:
                 change[k] = v
+
+        if change and self.target_type.model == 'subject':
+            from share.models import SubjectTaxonomy
```
Comment (chrisseto) [inline share/models/change.py]: Might be better to use `apps.get_model` here

File: share/util/__init__.py [hunk 11]
```diff
@@ -124,7 +124,7 @@ def __visit(self, node):
 
         self.__visiting.add(key)
         for k in self.__dependencies(node):
-            if k is not None:
+            if k is not None and k is not key:
```
Comment (chrisseto) [inline share/util/__init__.py]: What in what case is does `k == key`? Having a dependency on yourself should be considered an error.

File: tests/share/models/factories.py [hunk 12]
```diff
@@ -10,8 +10,18 @@
 from share import models
 
 
+class SourceFactory(DjangoModelFactory):
```
Comment (chrisseto) [inline tests/share/models/factories.py]: These factories already exist in `tests.factories.__init__`

File: share/models/meta.py [hunk 13]
```diff
@@ -50,28 +52,67 @@ class Disambiguation:
...
+        edge = node.related('central_synonym')
+        if edge and edge.related and edge.related.id == node.id:
+            graph.remove_edge(edge)
+
+    def save(self, *args, **kwargs):
+        if self.id is not None and self.parent is not None:
+            new_lineage = self.parent.lineage()
+            if self in new_lineage:
+                raise CyclicalTaxonomyError('Making {} a child of {} would cause a cycle!'.format(self, self.parent))
+        super().save(*args, **kwargs)
```
Comment (chrisseto) [inline share/models/meta.py]: nitpick: `return super()...`

File: share/models/meta.py [hunk 14]
```diff
@@ -50,28 +52,67 @@ class Disambiguation:
...
+    @classmethod
+    def normalize(cls, node, graph):
+        edge = node.related('central_synonym')
+        if edge and edge.related and edge.related.id == node.id:
+            graph.remove_edge(edge)
+
+    def save(self, *args, **kwargs):
+        if self.id is not None and self.parent is not None:
+            new_lineage = self.parent.lineage()
+            if self in new_lineage:
```
Comment (chrisseto) [inline share/models/meta.py]: Is this an appropriate place for this check to live?

File: bots/elasticsearch/bot.py (hunk 1 above)
Comment (aaxelb) [inline bots/elasticsearch/bot.py]:
```sql
EXPLAIN SELECT share_creativework.type, share_creativework.id, share_creativework.title, share_creativework.description, share_creativework.is_deleted, share_creativework.date_published, share_creativework.date_updated, share_creativework.free_to_read_type, share_creativework.free_to_read_date, share_creativework.rights, share_creativework.language, share_creativework.date_created, share_creativework.change_id, share_creativework.date_modified, share_creativework.version_id, share_creativework.same_as_id, share_creativework.same_as_version_id, share_creativework.extra_id, share_creativework.extra_version_id
FROM share_creativework
  LEFT OUTER JOIN share_throughsubjects ON (share_creativework.id = share_throughsubjects.creative_work_id)
  LEFT OUTER JOIN share_subject ON (share_throughsubjects.subject_id = share_subject.id)
WHERE (share_creativework.type IN ('share.software', 'share.project', 'share.patent', 'share.poster', 'share.conferencepaper', 'share.preprint', 'share.creativework', 'share.article', 'share.report', 'share.workingpaper', 'share.repository', 'share.retraction', 'share.publication', 'share.dissertation', 'share.book', 'share.registration', 'share.presentation', 'share.dataset', 'share.thesis')
       AND (
         share_creativework.date_modified > '2017-06-08T18:39:19.629395+00:00'
         OR share_subject.date_modified > '2017-06-08T18:39:19.629395+00:00'
         OR share_throughsubjects.date_modified > '2017-06-08T18:39:19.629395+00:00'
       )
)
```
(ugh typedmodels)
```
Hash Left Join  (cost=47.95..3994.97 rows=1 width=525)
  Hash Cond: (share_throughsubjects.subject_id = share_subject.id)
  Filter: ((share_creativework.date_modified > '2017-06-08 14:39:19.629395-04'::timestamp with time zone) OR (share_subject.date_modified > '2017-06-08 14:39:19.629395-04'::timestamp with time zone) OR (share_throughsubjects.date_modified > '2017-06-08 14:39:19.629395-04'::timestamp with time zone))
  ->  Merge Left Join  (cost=0.57..3495.88 rows=21257 width=537)
        Merge Cond: (share_creativework.id = share_throughsubjects.creative_work_id)
        ->  Index Scan using share_creativework_pkey on share_creativework  (cost=0.29..3082.48 rows=21257 width=525)
              Filter: ((type)::text = ANY ('{share.software,share.project,share.patent,share.poster,share.conferencepaper,share.preprint,share.creativework,share.article,share.report,share.workingpaper,share.repository,share.retraction,share.publication,share.dissertation,share.book,share.registration,share.presentation,share.dataset,share.thesis}'::text[]))
        ->  Index Scan using share_throughsubjects_creative_work_id_9bef7ff0 on share_throughsubjects  (cost=0.28..288.39 rows=6872 width=16)
  ->  Hash  (cost=32.17..32.17 rows=1217 width=12)
        ->  Seq Scan on share_subject  (cost=0.00..32.17 rows=1217 width=12)
```
Seems pretty decent? There is an index on `share_subject.date_modified`, for the record, but I guess postgres didn't feel like using it.

File: bots/elasticsearch/util.py (hunk 3 above)
Comment (aaxelb) [inline bots/elasticsearch/util.py]: Yeah, `concat_ws` conveniently ignores NULL values.

File: share/disambiguation.py (hunk 4 above)
Comment (aaxelb) [inline share/disambiguation.py]: Not if it's the system user initializing/updating the central taxonomy. Permissions are (somewhat hackily) checked when [accepting changes](https://github.com/CenterForOpenScience/SHARE/pull/674/files#diff-26baa8ae87e868e4bb3653a66e0d26f9R285).

File: share/disambiguation.py (hunk 5 above)
Comment (aaxelb) [inline share/disambiguation.py]: Similar to above. If it's, say, EngrXiv updating their custom taxonomy, nothing is wrong.

File: share/management/commands/disambiguate.py (hunk 6 above)
Comment (aaxelb) [inline share/management/commands/disambiguate.py]: Is that desired? I updated this command just so I could debug disambiguation, could update however you want.

File: share/migrations/0038_trust_system_user.py (hunk 7 above)
Comment (aaxelb) [inline share/migrations/0038_trust_system_user.py]: It doesn't have an icon, which I think is enough to exclude it from all the lists (is [this](https://github.com/CenterForOpenScience/SHARE/blob/develop/api/views/workflow.py#L45) the only place we list sources?)

File: share/migrations/0040_subject_provenance.py (hunk 8 above)
Comment (aaxelb) [inline share/migrations/0040_subject_provenance.py]: oh whoops, that's bad

File: share/util/__init__.py (hunk 11 above)
Comment (aaxelb) [inline share/util/__init__.py]: That's to support bepress subjects with `central_synonym` pointing at itself. It's replaced with `None` in `Subject.normalize`, but ChangeGraph sorts itself on init, before calling normalize, and will break here.

...that said, if the OSF side [is indeed switching to `None`](https://github.com/CenterForOpenScience/osf.io/pull/7243#discussion_r119704274), this is unnecessary.

File: tests/share/models/factories.py (hunk 12 above)
Comment (aaxelb) [inline tests/share/models/factories.py]: Oh hey, how about that.

File: share/models/meta.py (hunk 14 above)
Comment (aaxelb) [inline share/models/meta.py]: Preventing cycles seemed important enough to guarantee a check on every save... where do you think would make more sense?

**Comment by chrisseto** [comment]
@aaxelb when you have a moment, could you fix the merge conflicts?

**Comment by aaxelb** [comment]
@chrisseto done, sorry about that

**Comment by chrisseto** [comment]
@aaxelb, I made the conflicts. Sorry about that!

File: share/disambiguation.py [hunk 15]
```diff
@@ -144,6 +144,32 @@ def _instance_for_node(self, node):
...
+                try:
+                    return node.model.objects.get(name=name, central_synonym__isnull=True)
+                except node.model.DoesNotExist:
+                    logger.debug('No %s found with name "%s" in the central taxonomy', node.model, name)
+        else:
+            # Custom taxonomy
+            source = node.graph.source
+            if source and name:
+                try:
+                    return node.model.objects.get(name=name, taxonomy__name=source.long_title)
```
Comment (laurenbarker) [inline share/disambiguation.py]: Is basing the taxonomy name on the long_title fragile? If the source changes their long_title would we have to rename the taxonomy?

File: share/management/commands/disambiguate.py [hunk 16]
```diff
@@ -1,30 +1,22 @@
...
+from share.tasks import disambiguate
 
 
 class Command(BaseCommand):
 
     def add_arguments(self, parser):
-        parser.add_argument('source', type=str, help='The name of the provider to run')
         parser.add_argument('normalized', nargs='*', type=int, help='The id(*) of the normalized data to make changes for')
-        parser.add_argument('--all', action='store_true', help='Make changes for all normalized data for the source specified')
+        parser.add_argument('--source', type=str, help='Name of a Source, to disambiguate ALL its normalized data')
```
Comment (laurenbarker) [inline share/management/commands/disambiguate.py]: > to disambiguate ALL its normalized data

What does this mean?

File: tests/share/normalize/factories.py [hunk 17]
```diff
@@ -339,6 +343,12 @@ class ThroughTagsFactory(ShareObjectFactory):
     # creative_work = factory.SubFactory(AbstractCreativeWorkFactory)
 
 
+class ThroughSubjectsFactory(ShareObjectFactory):
```
Comment (laurenbarker) [inline tests/share/normalize/factories.py]: What's happening with this guy?

File: share/change.py [hunk 18]
```diff
@@ -250,11 +255,12 @@ def is_blank(self):
...
     @property
     def is_skippable(self):
-        if self.instance and self.graph.source and hasattr(self.instance, 'sources'):
-            new_source = not self.instance.sources.filter(source=self.graph.source).exists()
-        else:
-            new_source = False
-
+        new_source = False
+        if self.instance:
+            if self.type == 'subject' and self.instance.central_synonym is None and self.graph.source.user.username != settings.APPLICATION_USERNAME:
```
Comment (aaxelb) [inline share/change.py]: Yeah, if it's a Bepress subject (`central_synonym is None`) from a user besides `system`, this just silently ignores any changes.

File: share/disambiguation.py (hunk 15 above)
Comment (aaxelb) [inline share/disambiguation.py]: Hmm, good point. I guess it might be better to make SubjectTaxonomy one-to-one with Source? And the taxonomy wouldn't have a name, it would just use `taxonomy.source.long_title` when indexing or whatever.

That would actually simplify the "permission" logic: Each user can modify only their own source's taxonomy, and `system` is no longer a special case. Hm. I can do that, but it'll take several minutes...

File: share/management/commands/disambiguate.py (hunk 16 above)
Comment (aaxelb) [inline share/management/commands/disambiguate.py]: If you run `manage.py disambiguate --source=org.arxiv`, it'll disambiguate every single normalized datum from arxiv. I wanted to emphasize that it might be a significant amount of work.

Comment (laurenbarker) [inline share/management/commands/disambiguate.py]: Ohhhh for some reason I read it with a comma after ALL. Which I was thinking meant `manage.py disambiguate --source="normalized data"`. For my sake, could you rephrase that help text to something else, e.g. 'Name of a Source, to disambiguate ALL normalized data for that source'

File: tests/share/normalize/factories.py (hunk 17 above)
Comment (aaxelb) [inline tests/share/normalize/factories.py]: I copied `ThroughTagsFactory` three lines up, but it's part of the automagic `GraphContructor` [sic] which [looks for factories by name](https://github.com/aaxelb/SHARE/blob/3b059751dd229e9b0f1e218cf6022f4d4f093494/tests/share/normalize/factories.py#L126). The subfactories are commented out because these through-factories are only used indirectly, when both the work and tag/subject are given.

**Comment by aaxelb** [comment]
@laurenbarker [Data from the OSF](https://github.com/CenterForOpenScience/osf.io/pull/7243) will include each subject's parent (and its parent, etc. to the root). If a subject has null/missing `central_synonym` (or if `central_synonym` is a reference to itself), it's assumed to be a subject in the central taxonomy. Otherwise, the subject belongs to that source's custom taxonomy.
